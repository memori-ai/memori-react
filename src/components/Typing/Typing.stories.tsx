import React from 'react';
import { Meta, Story } from '@storybook/react';
import I18nWrapper from '../../I18nWrapper';
import Typing, { Props } from './Typing';

const meta: Meta = {
  title: 'Widget/Typing',
  component: Typing,
  argTypes: {
    useDefaultSentences: {
      control: {
        type: 'boolean',
      },
    },
    lang: {
      control: {
        type: 'select',
        options: ['en', 'it'],
      },
    },
  },
  parameters: {
    controls: { expanded: false },
  },
};

export default meta;

const Template: Story<Props> = args => (
  <I18nWrapper>
    <Typing {...args} />
  </I18nWrapper>
);
// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {};

export const WithLoadingTextIT = Template.bind({});
WithLoadingTextIT.args = {
  useDefaultSentences: true,
  lang: 'it',
};

export const WithLoadingTextEN = Template.bind({});
WithLoadingTextEN.args = {
  useDefaultSentences: true,
  lang: 'en',
};

export const WithCustomLoadingText = Template.bind({});
WithCustomLoadingText.args = {
  sentence: 'Chiedo agli unicorni cosa ne pensano...',
};

export const WithCustomLoadingTextList = Template.bind({});
WithCustomLoadingTextList.args = {
  sentences: {
    en: [
      {
        text: 'I am asking the unicorns what they think about it',
        delayAfter: 5,
      },
      {
        text: 'I am trying to understand what my cat is saying about it',
        delayAfter: 10,
      },
      {
        text: 'I am collecting the opinions of the people I know',
        delayAfter: 2,
      },
    ],
    it: [
      {
        text: 'Chiedo agli unicorni cosa ne pensano',
        delayAfter: 5,
      },
      {
        text: 'Sto cercando di capire cosa ne pensa il mio gatto',
        delayAfter: 10,
      },
      {
        text: 'Sto raccogliendo le opinioni delle persone che conosco',
        delayAfter: 2,
      },
    ],
  },
};

export const WithCustomLoadingTextListAndInitialDelay = Template.bind({});
WithCustomLoadingTextListAndInitialDelay.args = {
  sentences: {
    en: [
      {
        text: '',
        delayAfter: 10,
      },
      {
        text: 'I am asking the unicorns what they think about it',
        delayAfter: 5,
      },
      {
        text: 'I am trying to understand what my cat is saying about it',
        delayAfter: 10,
      },
      {
        text: 'I am collecting the opinions of the people I know',
        delayAfter: 2,
      },
    ],
    it: [
      {
        text: '',
        delayAfter: 10,
      },
      {
        text: 'Chiedo agli unicorni cosa ne pensano',
        delayAfter: 5,
      },
      {
        text: 'Sto cercando di capire cosa ne pensa il mio gatto',
        delayAfter: 10,
      },
      {
        text: 'Sto raccogliendo le opinioni delle persone che conosco',
        delayAfter: 2,
      },
    ],
  },
};
