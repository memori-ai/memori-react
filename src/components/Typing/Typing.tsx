import { useEffect, useState } from 'react';

const separator = ' ';
const defaultDelay = 3;
const defaultSentences = {
  en: [
    {
      delayAfter: defaultDelay,
      text: 'Generating an accurate and fancy response...',
    },
    {
      delayAfter: defaultDelay,
      text: 'Generating a response that will blow your mind...',
    },
    {
      delayAfter: defaultDelay,
      text: 'Generating a response that will make you smile...',
    },
    { delayAfter: defaultDelay, text: 'Thinking of a response...' },
    {
      delayAfter: defaultDelay,
      text: 'Thinking of a response that will make you smile...',
    },
    { delayAfter: defaultDelay, text: 'Gathering my thoughts...' },
    {
      delayAfter: defaultDelay,
      text: 'Gathering my thoughts to give you a response...',
    },
  ],
  it: [
    {
      delayAfter: defaultDelay,
      text: 'Sto generando una risposta accurata e fantasiosa...',
    },
    {
      delayAfter: defaultDelay,
      text: 'Sto generando una risposta che ti farà impazzire...',
    },
    {
      delayAfter: defaultDelay,
      text: 'Sto generando una risposta che ti farà sorridere...',
    },
    { delayAfter: defaultDelay, text: 'Sto pensando ad una risposta...' },
    {
      delayAfter: defaultDelay,
      text: 'Sto pensando ad una risposta che ti farà sorridere...',
    },
    { delayAfter: defaultDelay, text: 'Sto raccogliendo i miei pensieri...' },
    {
      delayAfter: defaultDelay,
      text: 'Sto raccogliendo i miei pensieri per darti una risposta...',
    },
  ],
};

export interface Props {
  useDefaultSentences?: boolean;
  lang?: 'en' | 'it';
  sentence?: string;
  sentences?: {
    [lang: string]: {
      /**
       * Sentence to show
       */
      text: string;
      /**
       * Seconds to wait after the sentence is completed
       */
      delayAfter: number;
    }[];
  };
}

const getSeparatorString = (seconds = defaultDelay) =>
  new Array(seconds * 20).fill(separator).join('');

const Typing = ({
  useDefaultSentences = false,
  lang = 'en',
  sentence,
  sentences,
}: Props) => {
  const [index, setIndex] = useState(0);
  const [text, setText] = useState(
    sentences?.[lang]?.length
      ? `${
          sentences[lang][0].text.endsWith('...') ||
          !sentences[lang][0].text.length
            ? sentences[lang][0].text
            : `${sentences[lang][0].text}...`
        }${getSeparatorString(sentences[lang][0].delayAfter)}`
      : sentence
      ? `${
          sentence.endsWith('...') || !sentence.length
            ? sentence
            : `${sentence}...`
        }${getSeparatorString()}`
      : ''
  );
  const [shownText, setShownText] = useState('');

  useEffect(() => {
    const interval = setInterval(() => {
      const letter = text[shownText.length];
      if (letter !== undefined && text.length > 0) {
        setShownText(prev => prev + letter);
      } else if (
        sentences?.[lang]?.length &&
        index < sentences[lang].length - 1
      ) {
        const nextIndex = index + 1;
        const sentence = sentences[lang][nextIndex];
        setText(
          `${
            sentence.text.endsWith('...') || !sentence.text.length
              ? sentence.text
              : `${sentence.text}...`
          }${getSeparatorString(sentence.delayAfter)}`
        );
        setShownText('');
        setIndex(nextIndex);
      } else if (!sentences && !sentence && useDefaultSentences) {
        const sentence =
          defaultSentences[lang][
            Math.floor(Math.random() * defaultSentences[lang].length)
          ];
        setText(`${sentence.text}${getSeparatorString(sentence.delayAfter)}`);
        setShownText('');
      }
    }, 50);

    return () => clearInterval(interval);
  });

  return (
    <div className="memori-chat--bubble">
      <div className="memori-chat--bubble-typing">
        <div id="wave">
          <span className="dot"></span>
          <span className="dot"></span>
          <span className="dot"></span>
        </div>
        {text.length > 0 && <p>{shownText}</p>}
      </div>
    </div>
  );
};

export default Typing;
