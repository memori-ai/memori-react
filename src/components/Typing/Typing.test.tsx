import { render } from '@testing-library/react';
import Typing from './Typing';

it('renders Typing unchanged', () => {
  const { container } = render(<Typing />);
  expect(container).toMatchSnapshot();
});

it('renders Typing with default loading text unchanged', () => {
  const { container } = render(<Typing useDefaultSentences lang="en" />);
  expect(container).toMatchSnapshot();
});

it('renders Typing with default italian loading text unchanged', () => {
  const { container } = render(<Typing useDefaultSentences lang="it" />);
  expect(container).toMatchSnapshot();
});

it('renders Typing with custom loading text unchanged', () => {
  const { container } = render(
    <Typing sentence="Chiedo agli unicorni cosa ne pensano..." />
  );
  expect(container).toMatchSnapshot();
});

it('renders Typing with custom loading text list unchanged', () => {
  const { container } = render(
    <Typing
      sentences={{
        en: [
          {
            text: 'I am asking the unicorns what they think about it',
            delayAfter: 5,
          },
          {
            text: 'I am trying to understand what my cat is saying about it',
            delayAfter: 10,
          },
          {
            text: 'I am collecting the opinions of the people I know',
            delayAfter: 2,
          },
        ],
        it: [
          {
            text: 'Chiedo agli unicorni cosa ne pensano',
            delayAfter: 5,
          },
          {
            text: 'Sto cercando di capire cosa ne pensa il mio gatto',
            delayAfter: 10,
          },
          {
            text: 'Sto raccogliendo le opinioni delle persone che conosco',
            delayAfter: 2,
          },
        ],
      }}
    />
  );
  expect(container).toMatchSnapshot();
});
