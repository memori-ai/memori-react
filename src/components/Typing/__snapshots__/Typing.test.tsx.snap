// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Typing unchanged 1`] = `
<div>
  <div
    class="memori-chat--bubble"
  >
    <div
      class="memori-chat--bubble-typing"
    >
      <div
        id="wave"
      >
        <span
          class="dot"
        />
        <span
          class="dot"
        />
        <span
          class="dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders Typing with custom loading text list unchanged 1`] = `
<div>
  <div
    class="memori-chat--bubble"
  >
    <div
      class="memori-chat--bubble-typing"
    >
      <div
        id="wave"
      >
        <span
          class="dot"
        />
        <span
          class="dot"
        />
        <span
          class="dot"
        />
      </div>
      <p />
    </div>
  </div>
</div>
`;

exports[`renders Typing with custom loading text unchanged 1`] = `
<div>
  <div
    class="memori-chat--bubble"
  >
    <div
      class="memori-chat--bubble-typing"
    >
      <div
        id="wave"
      >
        <span
          class="dot"
        />
        <span
          class="dot"
        />
        <span
          class="dot"
        />
      </div>
      <p />
    </div>
  </div>
</div>
`;

exports[`renders Typing with default italian loading text unchanged 1`] = `
<div>
  <div
    class="memori-chat--bubble"
  >
    <div
      class="memori-chat--bubble-typing"
    >
      <div
        id="wave"
      >
        <span
          class="dot"
        />
        <span
          class="dot"
        />
        <span
          class="dot"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders Typing with default loading text unchanged 1`] = `
<div>
  <div
    class="memori-chat--bubble"
  >
    <div
      class="memori-chat--bubble-typing"
    >
      <div
        id="wave"
      >
        <span
          class="dot"
        />
        <span
          class="dot"
        />
        <span
          class="dot"
        />
      </div>
    </div>
  </div>
</div>
`;
